Updates E2E Testing Documentation for local development (#1542)
google/fhir-data-pipes: doc/run-cloud-build-locally.md

# Running `cloudbuild.yaml` Locally with Docker

> `cloudbuild-local` is deprecated; however you can still iterate on the

Reviewer: ```suggestion
`cloud-build-local` is deprecated; however you can still iterate on the
```

`cloudbuild.yaml` steps by running them manually using Docker and the same
tooling that Cloud Build uses. This document provides a lightweight, repeatable
workflow for running the CI pipeline locally while being able to iterate
quickly.

## Prerequisites

- Docker Desktop (with `docker compose`) installed and running.
- A local JDK 17 + Maven 3.8.x for the Maven build steps.
- `python3` (for `synthea-uploader` unit tests) and `python -m unittest`
  available on `$PATH`.

> - Access to the workspace root (same directory that holds `cloudbuild.yaml`).

Reviewer: Should this simply be: "The GitHub repo is checked out locally" or something like that?

## Environment setup

1. Create an external Docker network named `cloudbuild`:

```bash
docker network create cloudbuild
```

2. From the repo root, export substitutions that Cloud Build normally injects:

```bash
export _TAG=local
export _REPOSITORY=fhir-analytics
```

3. If you need to mirror the GCP `gcloud` network configuration (Docker Compose

>    services talk to each other by name), run all commands inside this directory

Reviewer: This sentence is a little bit unclear; are you trying to say that all commands should be run from "the root of the repo" (which is fine and a good recommendation)? If yes, then what does have that to do with GCP network config and how docker compose services talk to each other?

SINK_SERVER_NAME=sink-server-jdbc SINK_SERVER_PORT=9002 docker compose -f docker/sink-compose.yml -p sink-server-jdbc up --force-recreate --remove-orphans -d
```

Wait for the servers to become ready and confirm via the command:

```bash
docker ps
```

You should get output similar to:

```
CONTAINER ID   IMAGE                     COMMAND                  CREATED              STATUS              PORTS                    NAMES
3a23504297d0   hapiproject/hapi:latest   "java --class-path /…"   About a minute ago   Up About a minute   0.0.0.0:9002->8080/tcp   sink-server-jdbc
c13ad0618e65   hapiproject/hapi:latest   "java --class-path /…"   2 minutes ago        Up 2 minutes        0.0.0.0:9001->8080/tcp   sink-server-search
1f301480d87f   hapiproject/hapi:latest   "java --class-path /…"   15 minutes ago       Up 15 minutes       0.0.0.0:8091->8080/tcp   hapi-server
815bc05a6005   postgres                  "docker-entrypoint.s…"   15 minutes ago       Up 15 minutes       0.0.0.0:5432->5432/tcp   hapi-fhir-db
```

### 2. Compile the Java modules

```bash
mvn --no-transfer-progress -e -T 2C install -Dlicense.skip=true -Dspotless.apply.skip=true
```

> ### 3. Upload sample synthetic data to HAPI stored in

Reviewer: "stored in ?" I think the title and first sentence below are connected (and need to be separated).

The uploader requires that you install the `uploader` module requirements first.
You can do this by running:

```bash
pip3 install -r ./synthea-hiv/uploader/requirements.txt
```

Please note, it is a good idea to first create a Python `virtualenv` before
running the above command to avoid conflicts with other Python packages you may
have installed globally. You can do this by running:

```bash
$ virtualenv -p python3.8 venv
$ . ./venv/bin/activate
```

If you had this already set up, just activate your existing virtualenv.

```bash
$ workon venv
```

Then, you can run the uploader script to upload the synthetic data to the FHIR
server.

For example, to upload to the HAPI FHIR server brought up in the previous step,
run:

```bash
python3 ./synthea-hiv/uploader/main.py HAPI http://localhost:9001/fhir \
--input_dir ./synthea-hiv/sample_data --cores 8
```

Depending on your machine, using too many cores may slow down your machine or
cause JDBC connection pool errors with the HAPI FHIR server. Reducing the number
of cores using the `--cores` flag should help at the cost of increasing the time
to upload the data.

### 4. Build the e2e controller/spark image

```bash

> docker build -t ${_REPOSITORY}/e2e-tests/controller-spark:${_TAG} -f e2e-tests/controller-spark/Dockerfile .

Reviewer: What I meant is that this image was not referenced later in this doc; I see that you have added that section now; thank you.

```

### 5. Build and run the batch pipeline

> Use the Maven-built image (or rebuild with `pipelines/batch/docker` if you

Reviewer: This is not very clear; what do you mean by "maven-built image"? Where did we create this docker image earlier?

change it):

```bash
cd pipelines/batch
docker build -t ${_REPOSITORY}/batch-pipeline:${_TAG} .
cd ../..
```

Run the FHIR-search job against HAPI:

```bash
docker run --rm \
  -e FHIR_SERVER_URL=http://localhost:9001/fhir \
  -e PARQUET_PATH=/workspace/e2e-tests/FHIR_SEARCH_HAPI \
  -e OUTPUT_PARQUET_VIEW_PATH=/workspace/e2e-tests/FHIR_SEARCH_HAPI/VIEWS_TIMESTAMP_1 \
  -e SINK_PATH=http://localhost:9001/fhir \
  -e SINK_USERNAME=hapi -e SINK_PASSWORD=hapi \

>   -v $(pwd):/workspace \

Reviewer: Consider adding a note about output directories under this path that might be overwritten.

Reviewer: I think in this case, it is `PARQUET_PATH` too, see [here](https://github.com/google/fhir-data-pipes/blob/547612dea431ebec8800fb99f9674150e06dce57/pipelines/batch/Dockerfile#L59).

  --network host  \
  ${_REPOSITORY}/batch-pipeline:${_TAG}
```

Repeat equivalent runs for JDBC mode, Bulk Export mode, and OpenMRS flows by
applying the same env vars/cloud build steps from `cloudbuild.yaml`.

### 6. Run e2e tests

Execute the HAPI search test:

```bash
  docker run --rm -e PARQUET_SUBDIR=FHIR_SEARCH_HAPI   \
    -e FHIR_JSON_SUBDIR=FHIR_SEARCH_HAPI_JSON  \
    -e SINK_SERVER=http://localhost:9001 \
    -e DOCKER_NETWORK=host   \
    -v $(pwd):/workspace   \
    --network host    \
    ${_REPOSITORY}/e2e-tests:${_TAG}
```

> or just the script directly:

Reviewer: This document that you have created is actually better than `cloud-build-local  --dryrun=false .` because even if we could use `cloud-build-local` we  still needed to replicate the steps locally to be able to debug (which is what this doc is). Please feel free if you want to use a better name for this, e.g., `debug-cloud-build-locally.md`.

```bash
./e2e-tests/pipeline_validation.sh e2e-tests/ FHIR_SEARCH_HAPI FHIR_SEARCH_HAPI_JSON http://localhost:9001/fhir
```

> Repeat for each mode (JDBC, Bulk Export, controller/spark, OpenMRS) using the

Reviewer: ditto (regarding "if required")
Also, please separate the `controller-spark` image from the rest and add a separate section for it. Its execution and debugging is different.

mvn --no-transfer-progress -e -T 2C install -Dlicense.skip=true -Dspotless.apply.skip=true
```

### 3. Upload sample synthetic data to HAPI

Upload sample synthetic data to HAPI stored in
[sample_data](https://github.com/google/fhir-data-pipes/blob/master/synthea-hiv/sample_data)
to the FHIR server that you brought up using the
[Synthea data uploader](https://github.com/google/fhir-data-pipes/blob/master/synthea-hiv/README.md#Uploader).

The uploader requires that you install the `uploader` module requirements.

It is a good idea to first create a Python `virtualenv` before running the
installation command to avoid conflicts with other Python packages you may have
installed globally.

If you use `virtualenv` then you can do it by running:

```bash
$ virtualenv -p python3 venv
$ . ./venv/bin/activate
```

or if you use the python standard library `venv` module, you can do it by
running:

```bash
$ python3 -m venv venv
$ source ./venv/bin/activate
```

> Then, you can then install the requirements with:

Reviewer: ```suggestion
Then, you can install the requirements with:
```

  -e FHIR_SERVER_URL=http://localhost:9001/fhir \
  -e PARQUET_PATH=/workspace/e2e-tests/BULK_EXPORT \
  -e OUTPUT_PARQUET_VIEW_PATH=/workspace/e2e-tests/BULK_EXPORT/VIEWS_TIMESTAMP_1 \
  -v $(pwd):/workspace \
  --network host \
  ${_REPOSITORY}/batch-pipeline:${_TAG}
```

### 2. Run e2e tests for Bulk Export mode

```bash
./e2e-tests/pipeline_validation.sh e2e-tests/ BULK_EXPORT BULK_EXPORT_FHIR_JSON
```

## Controller and Spark Mode

**Note:** Only run this section if debugging Controller and Spark mode;
otherwise skip.

### 1. Turn down sink servers

```bash
docker compose -f docker/sink-compose.yml -p sink-server-search down -v
docker compose -f docker/sink-compose.yml -p sink-server-jdbc down -v
```

### 2. Create views database

The `views` database is used for creating flat views from ViewDefinitions.

```bash
docker run --rm --network host postgres psql -U admin -d postgres -h localhost -p 5432 -c 'CREATE DATABASE views;'
```

### 3. Build the e2e controller/spark image

```bash
docker build -t ${_REPOSITORY}/e2e-tests/controller-spark:${_TAG} -f e2e-tests/controller-spark/Dockerfile .
```

### 4. Bring up controller and Spark containers

```bash

> PIPELINE_CONFIG=/workspace/docker/config DWH_ROOT=/workspace/e2e-tests/controller-spark/dwh docker compose -f docker/compose-controller-spark-sql-single.yaml up --force-recreate -d

Reviewer: Don't you need a volume mapping for `/workspace`?
nit: Also please  break the long line (didn't our default formatter catch this?).

Author: The bind mount is defined in the compose config file here - https://github.com/google/fhir-data-pipes/blob/547612dea431ebec8800fb99f9674150e06dce57/docker/compose-controller-spark-sql-single.yaml#L60-L62

Author: Don't think our default formatter(Spotless) can reformat these commands as appears in the `.md` files. We probably need to break these down manually.

```

### 5. Run e2e test for controller and Spark

```bash
docker run --rm -v $(pwd):/workspace ${_REPOSITORY}/e2e-tests/controller-spark:${_TAG}
```

> ### 6. Bring down controller and Spark containers

Reviewer: ```suggestion
### 6. Bring down controller and Spark containers
When you are done with debugging the controller-spark issues, you can bring those containers down.
```

Author: Done

```bash
docker compose -f docker/compose-controller-spark-sql-single.yaml down -v
```

## FHIR to FHIR Sync Mode

**Note:** Only run this section if debugging FHIR to FHIR sync mode; otherwise
skip.

### 1. Launch HAPI FHIR Sink Server for controller

```bash

> SINK_SERVER_NAME=sink-server-controller SINK_SERVER_PORT=9001 docker compose -f docker/sink-compose.yml -p sink-server up --force-recreate -d

Author: Done

```

> ### 2. Bring up pipeline controller for FHIR to FHIR sync

Reviewer: Please add a tip that people can use the controller from the previous step if they have it running (instead of bringing down/up).

```bash

> PIPELINE_CONFIG=/workspace/docker/config DWH_ROOT=/workspace/e2e-tests/controller-spark/dwh FHIRDATA_SINKFHIRSERVERURL=http://localhost:9001/fhir FHIRDATA_GENERATEPARQUETFILES=false FHIRDATA_CREATEHIVERESOURCETABLES=false FHIRDATA_CREATEPARQUETVIEWS=false FHIRDATA_SINKDBCONFIGPATH= docker compose -f docker/compose-controller-spark-sql-single.yaml up --force-recreate --no-deps -d pipeline-controller

Author: Done
